### PR TITLE
Protect endpoints

### PR DESCRIPTION
**Pull request by Uptaker**
This protects most endpoints so that only admins/users can view them. Also protects editing somebody else's posts/threads/account if they don't own it, unless they're an admin

### CODE REVIEW

File: backend/src/main/java/ee/tlu/forum/security/SecurityConfig.java [hunk 1]
```diff
@@ -46,14 +46,31 @@ protected void configure(HttpSecurity http) throws Exception {
         http.authorizeRequests().
                 antMatchers(AUTH_WHITELIST).permitAll();
 
-        // allow only certain roles access to these endpoints
-        http.authorizeRequests().antMatchers(GET, "/api/users/**").hasAnyAuthority("ROLE_USER");
-        http.authorizeRequests().antMatchers(GET, "/api/user/*").hasAnyAuthority("ROLE_USER");
-        http.authorizeRequests().antMatchers(POST, "/api/user/save/**").hasAnyAuthority("ROLE_ADMIN");
+        // UserController
+        http.authorizeRequests().antMatchers("/api/user/save/**").hasAnyAuthority("ROLE_ADMIN");
```
Comment (mariliisaar) [inline backend/src/main/java/ee/tlu/forum/security/SecurityConfig.java]: I'm guessing this is to edit user? What if it's the user editing their own profile and they're not an admin?

File: backend/src/main/java/ee/tlu/forum/service/ThreadService.java [hunk 2]
```diff
@@ -72,14 +75,17 @@ public void deleteThreadById(Long id) {
...
+    public Thread editThread(Thread thread, String token) {
         if (thread.getId() == null) {
             throw new BadRequestException("Field must contain a thread ID");
         }
         Optional<Thread> threadOptional = threadRepository.findById(thread.getId());
         if (threadOptional.isEmpty()) {
             throw new NotFoundException("Thread with ID " + thread.getId() + " does not exist");
         }
+
+        tokenHelper.hasRoleOrUsername(token, threadOptional.get().getAuthor().getUsername(), "ROLE_ADMIN");
```
Comment (mariliisaar) [inline backend/src/main/java/ee/tlu/forum/service/ThreadService.java]: Would this stop the author from editing their thread?

File: backend/src/main/java/ee/tlu/forum/service/ThreadServiceInterface.java [hunk 3]
```diff
@@ -6,9 +6,9 @@
 import java.util.List;
 
 public interface ThreadServiceInterface {
-    Thread createThread(AddNewThreadInput form);
+    Thread createThread(AddNewThreadInput form, String token);
     void deleteThreadById(Long id);
```
Comment (mariliisaar) [inline backend/src/main/java/ee/tlu/forum/service/ThreadServiceInterface.java:10]: Does this also need a token?

File: backend/src/main/java/ee/tlu/forum/service/UserServiceInterface.java [hunk 4]
```diff
@@ -6,7 +6,7 @@
 import java.util.List;
 
 public interface UserServiceInterface{
-    User editUser(User user);
+    User editUser(User user, String token);
     User registerUser(User user);
     Role saveRole(Role role);
```
Comment (mariliisaar) [inline backend/src/main/java/ee/tlu/forum/service/UserServiceInterface.java:11]: Token for this line and the next?

**Review by mariliisaar — APPROVED** [review]
Holy 🐄 (wanted to use mammoth but it doesn't seem to be working)!
Had a few questions mainly so I understand the code better - if you wouldn't mind explaining, I'd really appreciate it 🙂

File: backend/src/main/java/ee/tlu/forum/security/SecurityConfig.java (hunk 1 above)
Comment (Uptaker) [reply backend/src/main/java/ee/tlu/forum/security/SecurityConfig.java]: That's actually the wrong endpoint and you're right. I'll push a quickfix ASAP

File: backend/src/main/java/ee/tlu/forum/service/ThreadService.java (hunk 2 above)
Comment (Uptaker) [reply backend/src/main/java/ee/tlu/forum/service/ThreadService.java]: No, it stops other users from editing other people's threads

File: backend/src/main/java/ee/tlu/forum/service/ThreadServiceInterface.java (hunk 3 above)
Comment (Uptaker) [reply backend/src/main/java/ee/tlu/forum/service/ThreadServiceInterface.java:10]: Yes, as admins and users have different permissions but can access the same endpoint

File: backend/src/main/java/ee/tlu/forum/service/UserServiceInterface.java (hunk 4 above)
Comment (Uptaker) [reply backend/src/main/java/ee/tlu/forum/service/UserServiceInterface.java:11]: Yes, as admins and users have different permissions but can access the same endpoint

**Pull request merged**